### DISCUSSION TITLE
rofi -show window shows wrong workspaces

### DISCUSSION THREAD

**Discussion by tobiasrenkin** [i3/i3#5216] [Q&A]
Hi.

As the title says, rofi -show window shows the wrong workspace for all my open windows. It's not a huge issue but I am curious what is going on here.

Here is my i3 tree (printed using https://github.com/sciurus/i3tree)

```
Split (splith - horizontal)
	Workspace __i3_scratch (splith - none)
Split (splith - horizontal)
	Workspace 3: Browser (splith - horizontal)
		Split (tabbed - horizontal)
			Window New Discussion · i3/i3 — Mozilla Firefox
Split (splith - horizontal)
	Workspace 2: Editor (splith - horizontal)
		Split (tabbed - horizontal)
			Window config - Sublime Text
			Window ~/Dropbox/scripts/i3tree.pl
Split (splith - horizontal)
	Workspace 4: Mail (splith - none)

```
Rofi just uses wmctrl to get windows and workspaces. Here is what wmctrl gives me:

wmctrl -d

```
0  - DG: N/A  VP: 3520,0  WA: N/A  4: Mail
1  * DG: N/A  VP: 1600,0  WA: N/A  5: Org
2  - DG: N/A  VP: 0,300   WA: N/A  6: Media

```

wmctrl -l

```
0x03400020  1 XXX config - Sublime Text
0x00e001f2  0 XXX New Discussion · i3/i3 — Mozilla Firefox
0x0120000e  1 XXX wmctrl -l
```
As you can see, wmctrl gives a different window to workspace mapping than the i3 tree, and as a consequence the workspaces in rofi are wrong. In fact, wmctrl -d doesn't even show the correct list of open i3 workspaces, but other ones that are currently empty (workspace 5 and 6 instead of 2 and 3). 

How can that be? Any hints on where to look? I know this is more a rofi issue, but since rofi just takes the wmctrl output, which is somehow wrong as far as I can tell, I thought it better to open a discussion here. Hope it is not too off-topic.

Thanks,
Tobias

**Comment by Airblader** [comment]
The only way I can explain this is that the desktop hint atoms are not updated correctly.

Can you post the output of `xprop -root`? Also, `i3 --version`?

**Comment by tobiasrenkin** [comment]
Sure, here you go:

```
> i3 --version
i3 version 4.21-17-g812ec43d © 2009 Michael Stapelberg and contributors
```
```
> xprop -root
_NET_ACTIVE_WINDOW(WINDOW): window id # 0x120000e
AT_SPI_BUS(STRING) = "unix:abstract=/tmp/dbus-JZWtID55dt,guid=e6fac1a1651c610eb841a71b634fa7a4"
I3_SHMLOG_PATH(UTF8_STRING) = "/i3-log-1616"
I3_LOG_STREAM_SOCKET_PATH(UTF8_STRING) = "/run/user/1000/i3/log-stream-socket.1616"
I3_CONFIG_PATH(UTF8_STRING) = "/home/tobias/.config/i3/config"
I3_PID(CARDINAL) = 1616
I3_SOCKET_PATH(UTF8_STRING) = "/run/user/1000/i3/ipc-socket.1616"
_NET_CLIENT_LIST(WINDOW): window id # 0x3400020, 0xe001f2, 0x120000e, 0xc00002, 0xe0029b
_NET_CLIENT_LIST_STACKING(WINDOW): window id # 0xe0029b, 0xe001f2, 0x3400020, 0x120000e, 0xc00002
_NET_CURRENT_DESKTOP(CARDINAL) = 1
_NET_DESKTOP_NAMES(UTF8_STRING) = "4: Mail", "5: Org", "6: Media", ""
_NET_DESKTOP_VIEWPORT(CARDINAL) = 3520, 0, 1600, 0, 0, 300
_NET_NUMBER_OF_DESKTOPS(CARDINAL) = 3
_NET_SUPPORTED(ATOM) = _NET_SUPPORTED, _NET_SUPPORTING_WM_CHECK, _NET_WM_NAME, _NET_WM_VISIBLE_NAME, _NET_WM_MOVERESIZE, _NET_WM_STATE_STICKY, _NET_WM_STATE_FULLSCREEN, _NET_WM_STATE_DEMANDS_ATTENTION, _NET_WM_STATE_MODAL, _NET_WM_STATE_HIDDEN, _NET_WM_STATE_FOCUSED, _NET_WM_STATE, _NET_WM_WINDOW_TYPE, _NET_WM_WINDOW_TYPE_NORMAL, _NET_WM_WINDOW_TYPE_DOCK, _NET_WM_WINDOW_TYPE_DIALOG, _NET_WM_WINDOW_TYPE_UTILITY, _NET_WM_WINDOW_TYPE_TOOLBAR, _NET_WM_WINDOW_TYPE_SPLASH, _NET_WM_WINDOW_TYPE_MENU, _NET_WM_WINDOW_TYPE_DROPDOWN_MENU, _NET_WM_WINDOW_TYPE_POPUP_MENU, _NET_WM_WINDOW_TYPE_TOOLTIP, _NET_WM_WINDOW_TYPE_NOTIFICATION, _NET_WM_DESKTOP, _NET_WM_STRUT_PARTIAL, _NET_CLIENT_LIST, _NET_CLIENT_LIST_STACKING, _NET_CURRENT_DESKTOP, _NET_NUMBER_OF_DESKTOPS, _NET_DESKTOP_NAMES, _NET_DESKTOP_VIEWPORT, _NET_ACTIVE_WINDOW, _NET_CLOSE_WINDOW, _NET_MOVERESIZE_WINDOW
_NET_WM_NAME(UTF8_STRING) = "i3"
_NET_SUPPORTING_WM_CHECK(WINDOW): window id # 0x400047
SUPERSCRIPT_X(CARDINAL) = 
RESOURCE_MANAGER(STRING) = "*customization:\t-color\nXft.autohint:\tfalse\nXft.dpi:\t96\nXft.lcdfilter:\tlcddefault\nXft.hintstyle:\thintnone\nXft.rgba:\tnone\nXcursor.theme:\t\nXcursor.size:\t24\nXcursor.theme_core:\t1\n"
XFree86_DDC_EDID1_RAWDATA(INTEGER) = 0, -1, -1, -1, -1, -1, -1, 0, 9, -27, -37, 7, 0, 0, 0, 0, 1, 28, 1, 4, -91, 31, 17, 120, 2, 125, 80, -90, 87, 82, -97, 39, 18, 80, 84, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 58, 56, -128, -34, 112, 56, 40, 64, 48, 32, 54, 0, 53, -82, 16, 0, 0, 26, -5, 44, -128, -34, 112, 56, 40, 64, 48, 32, 54, 0, 53, -82, 16, 0, 0, 26, 0, 0, 0, -2, 0, 66, 79, 69, 32, 67, 81, 10, 32, 32, 32, 32, 32, 32, 0, 0, 0, -2, 0, 78, 69, 49, 52, 48, 70, 72, 77, 45, 78, 54, 49, 10, 0, 67
_GNOME_BACKGROUND_REPRESENTATIVE_COLORS(STRING) = "rgb(72,29,67)"
XKLAVIER_ALLOW_SECONDARY(INTEGER) = 0
GDK_VISUALS(INTEGER) = 1449, 1951
_XKB_RULES_NAMES(STRING) = "evdev", "pc104", "de,us", ",", ""
XFree86_has_VT(INTEGER) = 1
XFree86_VT(INTEGER) = 7
Xorg_Seat(STRING) = "seat0"
```

**Reply by Airblader** [reply]
Thanks. And to clarify, what i3bar shows you are workspaces 2: Editor, 3: Browser, and 4: Mail, right?

Do you know how you created these workspaces, IE just through normal usage, by loading a layout, ...? We'll need to figure out it how this situation came to be.

**Comment by tobiasrenkin** [comment]
Thanks for taking the time to think about this.

My i3bar shows the correct workspaces as they are in the tree.

Here are the lines in my i3 config defining the workspaces. I don't load any layouts etc. I changed the workspace names above to something readable because the symbols I use don't seem to work on github.

```

set $ws1 "1: "  
set $ws2 "2: " 
set $ws3 "3: " 
set $ws4 "4: " 
set $ws5 "5: "
set $ws6 "6: " 
set $ws7 "7: "
set $ws8 "8: "
set $ws9 "9: "
set $ws10 "10"

# HOME: HDMI-1 left, eDP-1 right
# Office: eDP-1 left, DP-1-1 center, DP-1-2 right

# assign to displays
workspace $ws1 output DP-1-1 HDMI-1 eDP-1
workspace $ws2 output DP-1-1 HDMI-1 eDP-1
workspace $ws3 output DP-1-2 HDMI-1 eDP-1
workspace $ws4 output eDP-1
workspace $ws5 output eDP-1
workspace $ws6 output eDP-1
workspace $ws7 output DP-1-2 HDMI-1 eDP-1
workspace $ws8 output DP-1-1 HDMI-1 eDP-1
workspace $ws9 output eDP-1

```

I also noticed some other odd things about the wmctrl -d output. First, the active workspace (the one with the *) is not correct. Second, the workspace names are not correct either. There is one workspace where I changed the symbol a while ago and it still shows the old one that is no longer used anywhere in my i3 config.

**Reply by Airblader** [reply]
It's clear now _what_ is happening / going wrong. What we don't yet know is _why_ that happens. I was only on my phone earlier, but now I can explain better:

What you see in i3bar and the i3tree output is the internal state of i3 queried from the IPC. However, neither wmctrl nor rofi know about this "proprietary" mechanism. Instead, they rely on the EWMH standard, which i3 does support. Basically, apart from its own IPC, i3 will also set atoms (which are like key-value pairs) on different windows. For example, on the root window, you saw

```
_NET_NUMBER_OF_DESKTOPS(CARDINAL) = 3
_NET_CURRENT_DESKTOP(CARDINAL) = 1
_NET_DESKTOP_NAMES(UTF8_STRING) = "4: Mail", "5: Org", "6: Media", ""
```

which tells tools like wmctrl how many desktops (that's the EWMH standard's name for what i3 calls workspace) there are, which one is currently active (its index, to be precise), and the name of those desktops.

You'll find similar atoms on the specific windows which tell wmctrl the index of the desktop on which the window manager claims the window is located.

These atoms are something that i3 actively needs to maintain and update as things change. Now, what's happening is that somehow, i3 fails to update this information, and wmctrl essentially operates in incorrect data.

This _does_ appear to be a bug in i3 from what I can tell right now. So I would please ask you to file this as a bug report rather than a discussion, feel free to link to the discussion for all the information you have already provided. What would be great is if you could also provide a log file, ideally on a new i3 session after doing as few things as possible to get into this situation, so that we can figure out why i3 fails to update this information. Thanks!

**Reply by tobiasrenkin** [reply]
OK, I will do that.

For the logs, should i just do i3-dump-log right after I log in, or do I need to enable logging in my dm somehow?

Let me know if there is anything else I can do to track this down, I really appreciate i3 and I am happy to help ironing out a bug.

**Reply by Airblader** [reply]
Thanks! You can find some information on getting a log file [here](https://i3wm.org/docs/debugging.html). 

> should i just do i3-dump-log right after I log in

The best scenario would be logging into a new session, setting up a simple case that shows the issue, and then pulling the log file, so that as few unrelated actions as possible occurred.

**Reply by tobiasrenkin** [reply]
I tried to reproduce the issue with the default config per the instructions in your link. And surprise, surprise, it doesn't seem to arise with the default config.

My config is pretty vanilla, but I will try to narrow down where it comes from before I file a bug report.

**Comment by tobiasrenkin** [comment]
OK, so the issue seems to come from xfsettingsd, which I start from my config with exec --no-startup-id xfsettingsd

- If I log into i3 without starting xfsettingsd, the problem is not there.
- If I start xfsettingsd after I logged in, things are still fine.
- If I create new workspaces after xfsettingsd has been started, the wmctrl workspace <-> window mapping starts being wrong.
- If I kill xfsettingsd, the mapping is still wrong.
- When I then restart i3 in place, the mapping is correct again.

I will upload some logs that document this. Should I still file this as a bug report, or leave it as a discussion?

**Reply by Airblader** [reply]
Ah, that's a nice find. Feel free to open the bug anyway -- in the end, it's the window manager's responsibility to make sure these atoms are correct, and we should figure out whether xfsettingsd is being naughty, or whether we are not handling something correctly.
